File: android/app/src/main/java/com/herobrawl/game/engine/Combat.kt
package com.herobrawl.game.engine

import com.herobrawl.game.data.Classes
import com.herobrawl.game.data.Factions
import com.herobrawl.game.data.Skills
import com.herobrawl.game.model.ClassId
import com.herobrawl.game.model.FactionId
import com.herobrawl.game.model.GameState
import com.herobrawl.game.model.HeroInstance
import com.herobrawl.game.model.Skill
import com.herobrawl.game.model.SkillTarget
import com.herobrawl.game.model.StatusEffect
import com.herobrawl.game.model.StatusKind
import kotlin.math.roundToInt
import kotlin.random.Random

data class BattleUnit(
    val uid: String,
    val side: Side,
    val slot: Int,
    val name: String,
    val emoji: String,
    val faction: FactionId,
    val classId: ClassId,
    val stars: Int,
    val gradientStart: Long,
    val gradientEnd: Long,
    val signatureColor: Long,
    val skills: List<Skill>,
    var attack: Int,
    var maxHp: Int,
    var hp: Int,
    var armor: Int,
    var speed: Int,
    var willpower: Int,
    var energy: Int = 0,
    val statuses: MutableList<StatusEffect> = mutableListOf(),
) {
    enum class Side { ALLY, ENEMY }
}

data class BattleLog(
    val kind: Kind,
    val text: String,
    val actor: String? = null,
    val target: String? = null,
    val amount: Int = 0,
) {
    enum class Kind { INFO, ATTACK, SKILL, HEAL, STATUS, DEATH, VICTORY, DEFEAT }
}

data class BattleResult(
    val winner: BattleUnit.Side,
    val turns: Int,
    val log: List<BattleLog>,
    val allyStart: List<BattleUnit>,
    val enemyStart: List<BattleUnit>,
)

object Combat {
    fun unitFromInstance(
        inst: HeroInstance, side: BattleUnit.Side, slot: Int,
        auraAtk: Double, auraHp: Double, auraSpd: Double,
    ): BattleUnit {
        val t = Stats.templateFor(inst)
        val s = Stats.compute(inst)
        val maxHp = (s.health * (1.0 + auraHp)).roundToInt()
        val atk = (s.attack * (1.0 + auraAtk)).roundToInt()
        val spd = (s.speed * (1.0 + auraSpd)).roundToInt()
        return BattleUnit(
            uid = "${side.name}-$slot-${inst.instanceId}",
            side = side, slot = slot,
            name = t.name, emoji = t.emoji,
            faction = t.faction, classId = t.heroClass,
            stars = inst.stars,
            gradientStart = t.portraitGradient[0],
            gradientEnd = t.portraitGradient[1],
            signatureColor = t.signatureColor,
            skills = t.skills,
            attack = atk, maxHp = maxHp, hp = maxHp,
            armor = s.armor, speed = spd, willpower = s.willpower,
            energy = if (t.heroClass == ClassId.MAGE) 30 else 0,
        )
    }

    fun buildAllyUnits(state: GameState): List<BattleUnit> {
        val insts = state.lineup.slots.mapNotNull { id ->
            id?.let { state.heroes.firstOrNull { h -> h.instanceId == it } }
        }
        val aura = Factions.lineupAura(insts.map { Stats.templateFor(it).faction })
        return state.lineup.slots.mapIndexedNotNull { slot, id ->
            val inst = id?.let { state.heroes.firstOrNull { h -> h.instanceId == it } } ?: return@mapIndexedNotNull null
            unitFromInstance(inst, BattleUnit.Side.ALLY, slot, aura.attack, aura.health, aura.speed)
        }
    }

    fun buildEnemyUnits(chapter: Int, stage: Int, rnd: Random = Random.Default): List<BattleUnit> {
        val progress = (chapter - 1) * 10 + (stage - 1)
        val power = 1.0 + progress * 0.18
        val slotPicks = listOf(ClassId.GUARDIAN, ClassId.GUARDIAN, ClassId.BERSERKER, ClassId.MAGE, ClassId.RANGER)
        val names = listOf("Rogue Sellsword", "Cursed Knight", "Wildfire Witch",
            "Night Stalker", "Ironclad", "Lost Pilgrim")
        val emojiFor = mapOf(
            ClassId.GUARDIAN to "🗿", ClassId.BERSERKER to "💢", ClassId.MAGE to "🌀",
            ClassId.ASSASSIN to "🕸️", ClassId.RANGER to "🎯", ClassId.CLERIC to "🕯️",
        )
        val factions = FactionId.values()

        return (0 until 5).map { slot ->
            val cls = slotPicks[slot]
            val faction = factions[(slot * 31 + chapter * 17 + stage * 7 + rnd.nextInt(0, 1_000_000)) % factions.size]
            val base = Classes.all.getValue(cls).baseStats
            val stars = minOf(5, 2 + progress / 8)
            val mul = power * (0.85 + stars * 0.1)
            val maxHp = (base.health * mul).roundToInt()
            BattleUnit(
                uid = "enemy-$slot-$progress-${rnd.nextInt()}",
                side = BattleUnit.Side.ENEMY,
                slot = slot,
                name = "${names[slot % names.size]} ${progress + 1}",
                emoji = emojiFor.getValue(cls),
                faction = faction,
                classId = cls,
                stars = stars,
                gradientStart = 0xFF2A2A36,
                gradientEnd = 0xFF4A4A5A,
                signatureColor = 0xFF8A8A9A,
                skills = Skills.defaultFor(cls),
                attack = (base.attack * mul).roundToInt(),
                maxHp = maxHp,
                hp = maxHp,
                armor = (base.armor * mul * 0.8).roundToInt(),
                speed = base.speed,
                willpower = base.willpower,
            )
        }
    }

    fun simulate(
        allies: List<BattleUnit>, enemies: List<BattleUnit>,
        rnd: Random = Random.Default, maxTurns: Int = 40,
    ): BattleResult {
        val A = allies.map { it.copy(statuses = it.statuses.toMutableList(), skills = it.skills.toList()) }
        val E = enemies.map { it.copy(statuses = it.statuses.toMutableList(), skills = it.skills.toList()) }
        val log = mutableListOf<BattleLog>()
        log += BattleLog(BattleLog.Kind.INFO, "Battle begins!")
        var turn = 0
        val allyStart = A.map { it.copy(statuses = mutableListOf()) }
        val enemyStart = E.map { it.copy(statuses = mutableListOf()) }

        while (turn < maxTurns) {
            turn++
            val order = (A + E).filter { it.hp > 0 }.sortedByDescending { it.speed }
            for (actor in order) {
                if (actor.hp <= 0) continue
                if (tickStatuses(actor, log)) continue
                val opponents = if (actor.side == BattleUnit.Side.ALLY) E else A
                if (opponents.all { it.hp <= 0 }) break

                actor.energy = minOf(100, actor.energy + 25 + if (actor.classId == ClassId.MAGE) 10 else 0)
                val active = actor.skills.firstOrNull { it.active }
                if (actor.energy >= 100 && active != null) {
                    actor.energy = 0
                    executeSkill(actor, active, A, E, log, rnd)
                } else {
                    basicAttack(actor, opponents, log, rnd)
                }
                if (A.all { it.hp <= 0 } || E.all { it.hp <= 0 }) break
            }
            for (u in A + E) if (u.hp > 0) applyDots(u, log)
            if (A.all { it.hp <= 0 } || E.all { it.hp <= 0 }) break
        }

        val winner = if (A.any { it.hp > 0 }) BattleUnit.Side.ALLY else BattleUnit.Side.ENEMY
        log += BattleLog(
            if (winner == BattleUnit.Side.ALLY) BattleLog.Kind.VICTORY else BattleLog.Kind.DEFEAT,
            if (winner == BattleUnit.Side.ALLY) "Victory!" else "Defeat."
        )
        return BattleResult(winner, turn, log.toList(), allyStart, enemyStart)
    }

    private fun tickStatuses(u: BattleUnit, log: MutableList<BattleLog>): Boolean {
        var stunned = false
        val keep = mutableListOf<StatusEffect>()
        for (s in u.statuses) {
            if (s.kind == StatusKind.STUN && s.duration > 0) {
                stunned = true
                log += BattleLog(BattleLog.Kind.STATUS, "${u.name} is stunned.", u.uid)
            }
            val d = s.duration - 1
            if (d > 0) keep += s.copy(duration = d)
        }
        u.statuses.clear(); u.statuses.addAll(keep)
        return stunned
    }

    private fun applyDots(u: BattleUnit, log: MutableList<BattleLog>) {
        for (s in u.statuses) when (s.kind) {
            StatusKind.BURN, StatusKind.POISON -> {
                val dmg = (u.maxHp * s.magnitude).roundToInt()
                u.hp = maxOf(0, u.hp - dmg)
                log += BattleLog(BattleLog.Kind.STATUS, "${u.name} suffers $dmg ${s.kind.name.lowercase()} damage.", u.uid, amount = dmg)
                if (u.hp == 0) log += BattleLog(BattleLog.Kind.DEATH, "${u.name} falls.", u.uid)
            }
            StatusKind.REGEN -> {
                val heal = (u.maxHp * s.magnitude).roundToInt()
                u.hp = minOf(u.maxHp, u.hp + heal)
                log += BattleLog(BattleLog.Kind.HEAL, "${u.name} regenerates $heal HP.", u.uid, amount = heal)
            }
            else -> Unit
        }
    }

    private fun pickTarget(actor: BattleUnit, opponents: List<BattleUnit>): BattleUnit? {
        val alive = opponents.filter { it.hp > 0 }
        if (alive.isEmpty()) return null
        if (actor.classId == ClassId.ASSASSIN) return alive.minBy { it.hp }
        if (actor.classId == ClassId.RANGER || actor.classId == ClassId.MAGE) {
            val back = alive.filter { it.slot >= 3 }
            if (back.isNotEmpty()) return back.first()
        }
        val front = alive.filter { it.slot <= 2 }
        return if (front.isNotEmpty()) front.first() else alive.first()
    }

    private fun basicAttack(actor: BattleUnit, opponents: List<BattleUnit>, log: MutableList<BattleLog>, rnd: Random) {
        val t = pickTarget(actor, opponents) ?: return
        val dmg = computeDamage(actor, t, 1.0, rnd)
        t.hp = maxOf(0, t.hp - dmg)
        log += BattleLog(BattleLog.Kind.ATTACK, "${actor.name} attacks ${t.name} for $dmg.", actor.uid, t.uid, dmg)
        if (t.hp == 0) log += BattleLog(BattleLog.Kind.DEATH, "${t.name} falls.", t.uid)
    }

    private fun executeSkill(
        actor: BattleUnit, skill: Skill,
        allies: List<BattleUnit>, enemies: List<BattleUnit>,
        log: MutableList<BattleLog>, rnd: Random,
    ) {
        val opponents = if (actor.side == BattleUnit.Side.ALLY) enemies else allies
        val friends = if (actor.side == BattleUnit.Side.ALLY) allies else enemies
        log += BattleLog(BattleLog.Kind.SKILL, "${actor.name} unleashes ${skill.name}!", actor.uid)
        val power = skill.power
        when {
            skill.target == SkillTarget.ALL_ENEMIES || skill.aoe -> {
                for (t in opponents.filter { it.hp > 0 }) {
                    val dmg = computeDamage(actor, t, power, rnd)
                    t.hp = maxOf(0, t.hp - dmg)
                    log += BattleLog(BattleLog.Kind.ATTACK, "→ hits ${t.name} for $dmg.", actor.uid, t.uid, dmg)
                    skill.status?.let { t.statuses += it.copy() }
                    if (t.hp == 0) log += BattleLog(BattleLog.Kind.DEATH, "${t.name} falls.", t.uid)
                }
            }
            skill.target == SkillTarget.ALL_ALLIES -> {
                val heal = (actor.attack * power).roundToInt()
                for (f in friends.filter { it.hp > 0 }) {
                    f.hp = minOf(f.maxHp, f.hp + heal)
                    log += BattleLog(BattleLog.Kind.HEAL, "→ restores $heal HP to ${f.name}.", actor.uid, f.uid, heal)
                    skill.status?.let { f.statuses += it.copy() }
                }
            }
            else -> {
                val t = pickTarget(actor, opponents) ?: return
                val dmg = computeDamage(actor, t, power, rnd)
                t.hp = maxOf(0, t.hp - dmg)
                log += BattleLog(BattleLog.Kind.ATTACK, "→ strikes ${t.name} for $dmg.", actor.uid, t.uid, dmg)
                skill.status?.let { t.statuses += it.copy() }
                if (t.hp == 0) log += BattleLog(BattleLog.Kind.DEATH, "${t.name} falls.", t.uid)
                if (skill.status?.kind == StatusKind.SHIELD) actor.statuses += skill.status.copy()
            }
        }
    }

    private fun computeDamage(
        attacker: BattleUnit, defender: BattleUnit, powerMul: Double, rnd: Random,
    ): Int {
        val adv = Factions.advantage(attacker.faction, defender.faction)
        var dmg = attacker.attack * powerMul
        if (attacker.classId == ClassId.BERSERKER) {
            val missing = 1.0 - attacker.hp.toDouble() / attacker.maxHp.coerceAtLeast(1)
            dmg *= 1.0 + missing
        }
        if (attacker.classId == ClassId.RANGER) dmg *= 1.15
        val armor = defender.armor * (if (attacker.classId == ClassId.RANGER) 0.7 else 1.0)
        val mitigation = armor / (armor + 500.0)
        dmg *= 1.0 - mitigation
        dmg *= 1.0 + adv
        val critChance = if (attacker.classId == ClassId.ASSASSIN &&
            defender.hp < defender.maxHp * 0.5) 0.5 else 0.2
        val crit = rnd.nextDouble() < critChance
        if (crit) dmg *= 1.8
        if (defender.classId == ClassId.GUARDIAN &&
            (attacker.classId == ClassId.BERSERKER || attacker.classId == ClassId.ASSASSIN ||
                attacker.classId == ClassId.GUARDIAN)) dmg *= 0.85
        val shield = defender.statuses.firstOrNull { it.kind == StatusKind.SHIELD }
        if (shield != null) dmg *= 1.0 - minOf(0.4, shield.magnitude)
        return maxOf(1, dmg.roundToInt())
    }

    data class Rewards(
        val gold: Long, val spirit: Long, val xp: Int, val brawlPoints: Int,
        val firstClearGems: Long = 0,
    )

    fun campaignRewards(chapter: Int, stage: Int, firstClear: Boolean): Rewards {
        val progress = (chapter - 1) * 10 + (stage - 1)
        return Rewards(
            gold = 800L + progress * 120,
            spirit = 30L + progress * 6,
            xp = 40 + progress * 8,
            brawlPoints = 20 + progress * 3,
            firstClearGems = if (firstClear) 20L else 0L,
        )
    }
}

File: src/game/combat.ts
import type {
  BattleLog,
  BattleResult,
  BattleUnit,
  FactionId,
  GameState,
  HeroInstance,
  Skill,
  StatusEffect,
} from "../types";
import { CLASSES } from "./classes";
import { factionAdvantage, lineupAura } from "./factions";
import { CLASS_ACTIVE } from "./skills";
import { computeStats, templateFor } from "./stats";

// Build a battle unit from a hero instance.
export function unitFromInstance(
  inst: HeroInstance,
  side: "ally" | "enemy",
  slot: number,
  auraBonus?: { attack: number; health: number; speed: number }
): BattleUnit {
  const t = templateFor(inst);
  const stats = computeStats(inst);
  const mul = auraBonus ?? { attack: 0, health: 0, speed: 0 };
  const maxHp = Math.round(stats.health * (1 + mul.health));
  return {
    uid: `${side}-${slot}-${inst.instanceId}`,
    side,
    slot,
    name: t.name,
    emoji: t.emoji,
    faction: t.faction,
    class: t.class,
    stars: inst.stars,
    stats: {
      ...stats,
      attack: Math.round(stats.attack * (1 + mul.attack)),
      speed: Math.round(stats.speed * (1 + mul.speed)),
    },
    hp: maxHp,
    maxHp,
    energy: t.class === "mage" ? 30 : 0,
    skills: t.skills,
    statuses: [],
    portraitGradient: t.portraitGradient,
    signatureColor: t.signatureColor,
  };
}

export function buildAllyUnits(state: GameState): BattleUnit[] {
  const instIds = state.lineup.slots.filter(Boolean) as string[];
  if (!instIds.length) return [];
  const insts = instIds
    .map((id) => state.heroes.find((h) => h.instanceId === id))
    .filter(Boolean) as HeroInstance[];
  const factions: FactionId[] = insts.map((i) => templateFor(i).faction);
  const aura = lineupAura(factions);
  return state.lineup.slots.map((id, slot) => {
    if (!id) return null;
    const inst = state.heroes.find((h) => h.instanceId === id);
    if (!inst) return null;
    return unitFromInstance(inst, "ally", slot, aura.bonus);
  }).filter(Boolean) as BattleUnit[];
}

// Build scaling enemy units.
export function buildEnemyUnits(chapter: number, stage: number, seed = Math.random()): BattleUnit[] {
  const progress = (chapter - 1) * 10 + (stage - 1);
  const power = 1 + progress * 0.18;

  const archetypes: (keyof typeof CLASSES)[] = [
    "guardian",
    "berserker",
    "mage",
    "assassin",
    "ranger",
    "cleric",
  ];
  const factions: FactionId[] = [
    "vanguard",
    "horde",
    "wildwood",
    "arcane",
    "radiance",
    "abyss",
  ];

  const units: BattleUnit[] = [];
  const slotPicks: (keyof typeof CLASSES)[] = [
    "guardian",
    "guardian",
    "berserker",
    "mage",
    "ranger",
  ];

  for (let slot = 0; slot < 5; slot++) {
    const seeded = (seed * 1000 + slot * 17 + progress * 31) % 1;
    const cls = slotPicks[slot] ?? archetypes[Math.floor(seeded * archetypes.length)];
    const faction = factions[Math.floor(((seeded + slot * 0.13 + chapter * 0.07) % 1) * factions.length)];
    const base = CLASSES[cls].baseStats;

    const stars = Math.min(5, 2 + Math.floor(progress / 8)) as BattleUnit["stars"];
    const statMul = power * (0.85 + stars * 0.1);
    const maxHp = Math.round(base.health * statMul);
    const name = ENEMY_NAMES[slot % ENEMY_NAMES.length] + " " + (progress + 1);

    units.push({
      uid: `enemy-${slot}-${progress}`,
      side: "enemy",
      slot,
      name,
      emoji: ENEMY_EMOJI[cls],
      faction,
      class: cls,
      stars,
      stats: {
        attack: Math.round(base.attack * statMul),
        health: maxHp,
        armor: Math.round(base.armor * statMul * 0.8),
        speed: base.speed,
        willpower: base.willpower,
      },
      hp: maxHp,
      maxHp,
      energy: 0,
      skills: [CLASS_ACTIVE[cls]],
      statuses: [],
      portraitGradient: ["#2a2a36", "#4a4a5a"],
      signatureColor: "#8a8a9a",
    });
  }

  return units;
}

const ENEMY_NAMES = [
  "Rogue Sellsword",
  "Cursed Knight",
  "Wildfire Witch",
  "Night Stalker",
  "Ironclad",
  "Lost Pilgrim",
];
const ENEMY_EMOJI: Record<keyof typeof CLASSES, string> = {
  guardian: "🗿",
  berserker: "💢",
  mage: "🌀",
  assassin: "🕸️",
  ranger: "🎯",
  cleric: "🕯️",
};

// --- Combat simulator ---------------------------------------------------

export interface BattleOptions {
  manualBurst?: boolean; // if true, ally skills fire when energy full next chance
  seed?: number;
}

export function simulateBattle(
  allies: BattleUnit[],
  enemies: BattleUnit[],
  opts: BattleOptions = {}
): BattleResult {
  void opts;
  // Deep-copy units so callers' references remain clean.
  const A = allies.map((u) => cloneUnit(u));
  const E = enemies.map((u) => cloneUnit(u));
  const log: BattleLog[] = [];

  const pushLog = (entry: Omit<BattleLog, "at">) =>
    log.push({ ...entry, at: performance.now() });

  pushLog({ kind: "info", text: "Battle begins!" });

  let turn = 0;
  const maxTurns = 40;

  while (turn < maxTurns) {
    turn++;
    // Build turn order by speed, all alive units.
    const all = [...A, ...E].filter((u) => u.hp > 0);
    all.sort((a, b) => b.stats.speed - a.stats.speed);

    for (const actor of all) {
      if (actor.hp <= 0) continue;
      if (tickStatuses(actor, pushLog)) continue; // stunned

      const opponents = actor.side === "ally" ? E : A;
      if (opponents.every((u) => u.hp <= 0)) break;

      // Charge energy
      actor.energy = Math.min(100, actor.energy + 25 + (actor.class === "mage" ? 10 : 0));

      // Fire active skill when energy >= 100
      if (actor.energy >= 100 && actor.skills.some((s) => s.kind === "active")) {
        const skill = actor.skills.find((s) => s.kind === "active")!;
        actor.energy = 0;
        executeSkill(actor, skill, A, E, pushLog);
      } else {
        basicAttack(actor, opponents, pushLog);
      }

      if (A.every((u) => u.hp <= 0) || E.every((u) => u.hp <= 0)) break;
    }

    // End-of-turn statuses (burn/poison/regen)
    for (const u of [...A, ...E]) if (u.hp > 0) applyDots(u, pushLog);

    if (A.every((u) => u.hp <= 0) || E.every((u) => u.hp <= 0)) break;
  }

  const winner: "ally" | "enemy" = A.some((u) => u.hp > 0) ? "ally" : "enemy";
  pushLog({
    kind: winner === "ally" ? "victory" : "defeat",
    text: winner === "ally" ? "Victory!" : "Defeat.",
  });

  return { winner, turns: turn, log, survivors: [...A, ...E].filter((u) => u.hp > 0) };
}

function cloneUnit(u: BattleUnit): BattleUnit {
  return {
    ...u,
    stats: { ...u.stats },
    statuses: u.statuses.map((s) => ({ ...s })),
    skills: u.skills.map((s) => ({ ...s })),
  };
}

function tickStatuses(u: BattleUnit, log: (e: Omit<BattleLog, "at">) => void): boolean {
  let stunned = false;
  const next: StatusEffect[] = [];
  for (const s of u.statuses) {
    if (s.kind === "stun" && s.duration > 0) {
      stunned = true;
      log({ kind: "status", text: `${u.name} is stunned.`, actor: u.uid });
    }
    const d = s.duration - 1;
    if (d > 0) next.push({ ...s, duration: d });
  }
  u.statuses = next;
  return stunned;
}

function applyDots(u: BattleUnit, log: (e: Omit<BattleLog, "at">) => void) {
  for (const s of u.statuses) {
    if (s.kind === "burn" || s.kind === "poison") {
      const dmg = Math.round(u.maxHp * s.magnitude);
      u.hp = Math.max(0, u.hp - dmg);
      log({
        kind: "status",
        text: `${u.name} suffers ${dmg} ${s.kind} damage.`,
        actor: u.uid,
        amount: dmg,
      });
      if (u.hp <= 0) {
        log({ kind: "death", text: `${u.name} falls.`, actor: u.uid });
      }
    } else if (s.kind === "regen") {
      const heal = Math.round(u.maxHp * s.magnitude);
      u.hp = Math.min(u.maxHp, u.hp + heal);
      log({
        kind: "heal",
        text: `${u.name} regenerates ${heal} HP.`,
        actor: u.uid,
        amount: heal,
      });
    }
  }
}

function pickTarget(actor: BattleUnit, opponents: BattleUnit[]): BattleUnit | undefined {
  const alive = opponents.filter((u) => u.hp > 0);
  if (!alive.length) return undefined;

  // Assassins target lowest HP, rangers/mages prefer backline, others target frontline
  if (actor.class === "assassin") {
    return [...alive].sort((a, b) => a.hp - b.hp)[0];
  }
  if (actor.class === "ranger" || actor.class === "mage") {
    const back = alive.filter((u) => u.slot >= 3);
    if (back.length) return back[0];
  }
  const front = alive.filter((u) => u.slot <= 2);
  if (front.length) return front[0];
  return alive[0];
}

function basicAttack(
  actor: BattleUnit,
  opponents: BattleUnit[],
  log: (e: Omit<BattleLog, "at">) => void
) {
  const target = pickTarget(actor, opponents);
  if (!target) return;
  const { damage } = computeDamage(actor, target, 1.0);
  target.hp = Math.max(0, target.hp - damage);
  log({
    kind: "attack",
    text: `${actor.name} attacks ${target.name} for ${damage}.`,
    actor: actor.uid,
    target: target.uid,
    amount: damage,
  });
  if (target.hp <= 0) log({ kind: "death", text: `${target.name} falls.`, actor: target.uid });
}

function executeSkill(
  actor: BattleUnit,
  skill: Skill,
  allies: BattleUnit[],
  enemies: BattleUnit[],
  log: (e: Omit<BattleLog, "at">) => void
) {
  const opponents = actor.side === "ally" ? enemies : allies;
  const friends = actor.side === "ally" ? allies : enemies;

  log({
    kind: "skill",
    text: `${actor.name} unleashes ${skill.name}!`,
    actor: actor.uid,
  });

  const power = skill.power ?? 1;

  if (skill.target === "all-enemies" || skill.aoe) {
    for (const t of opponents.filter((u) => u.hp > 0)) {
      const { damage } = computeDamage(actor, t, power);
      t.hp = Math.max(0, t.hp - damage);
      log({
        kind: "attack",
        text: `→ hits ${t.name} for ${damage}.`,
        actor: actor.uid,
        target: t.uid,
        amount: damage,
      });
      if (skill.status) t.statuses.push({ ...skill.status });
      if (t.hp <= 0) log({ kind: "death", text: `${t.name} falls.`, actor: t.uid });
    }
  } else if (skill.target === "all-allies") {
    const heal = Math.round(actor.stats.attack * power);
    for (const f of friends.filter((u) => u.hp > 0)) {
      f.hp = Math.min(f.maxHp, f.hp + heal);
      log({
        kind: "heal",
        text: `→ restores ${heal} HP to ${f.name}.`,
        actor: actor.uid,
        target: f.uid,
        amount: heal,
      });
      if (skill.status) f.statuses.push({ ...skill.status });
    }
  } else {
    // single enemy
    const target = pickTarget(actor, opponents);
    if (!target) return;
    const { damage } = computeDamage(actor, target, power);
    target.hp = Math.max(0, target.hp - damage);
    log({
      kind: "attack",
      text: `→ strikes ${target.name} for ${damage}.`,
      actor: actor.uid,
      target: target.uid,
      amount: damage,
    });
    if (skill.status) target.statuses.push({ ...skill.status });
    if (target.hp <= 0) log({ kind: "death", text: `${target.name} falls.`, actor: target.uid });
    // Self-shield for guardians (Bulwark Strike secondary effect)
    if (skill.status?.kind === "shield") {
      actor.statuses.push({ ...skill.status });
    }
  }
}

function computeDamage(
  attacker: BattleUnit,
  defender: BattleUnit,
  powerMul: number
): { damage: number; crit: boolean } {
  const adv = factionAdvantage(attacker.faction, defender.faction);
  let dmg = attacker.stats.attack * powerMul;

  // Class passives
  if (attacker.class === "berserker") {
    const missing = 1 - attacker.hp / Math.max(1, attacker.maxHp);
    dmg *= 1 + missing; // Bloodrage
  }
  if (attacker.class === "ranger") {
    // ignore 30% armor
    dmg *= 1.15;
  }

  // Armor mitigation
  const armor = defender.stats.armor * (attacker.class === "ranger" ? 0.7 : 1);
  const mitigation = armor / (armor + 500);
  dmg *= 1 - mitigation;

  // Faction advantage
  dmg *= 1 + adv.damage;

  // Crit
  const critChance = attacker.class === "assassin" && defender.hp < defender.maxHp * 0.5
    ? 0.5
    : 0.2;
  const crit = Math.random() < critChance;
  if (crit) dmg *= 1.8;

  // Guardian passive: 15% less from melee
  if (defender.class === "guardian" && (attacker.class === "berserker" || attacker.class === "assassin" || attacker.class === "guardian")) {
    dmg *= 0.85;
  }

  // Defender shield
  const shield = defender.statuses.find((s) => s.kind === "shield");
  if (shield) {
    dmg *= 1 - Math.min(0.4, shield.magnitude);
  }

  // Cleric Radiant Aegis
  const sameSide = attacker.side !== defender.side;
  if (sameSide) {
    const cleric = (defender.side === "ally")
      ? allyHasCleric(defender, dmg)
      : false;
    if (cleric && defender.hp / defender.maxHp < 0.3) dmg *= 0.75;
  }

  return { damage: Math.max(1, Math.round(dmg)), crit };
}

function allyHasCleric(u: BattleUnit, dmg: number) {
  void u;
  void dmg;
  // Simplified: assume absence; full implementation would scan friends.
  return false;
}

// --- Rewards -------------------------------------------------------------

export interface BattleRewards {
  gold: number;
  spirit: number;
  xp: number;
  brawlPoints: number;
  firstClear?: { gems: number };
}

export function rewardsForCampaign(chapter: number, stage: number, firstClear: boolean): BattleRewards {
  const progress = (chapter - 1) * 10 + (stage - 1);
  const baseGold = 800 + progress * 120;
  const baseSpirit = 30 + progress * 6;
  const xp = 40 + progress * 8;
  return {
    gold: baseGold,
    spirit: baseSpirit,
    xp,
    brawlPoints: 20 + progress * 3,
    firstClear: firstClear ? { gems: 20 } : undefined,
  };
}
